Replace direct `Command` by crates in the build script

Currently, the build script contains two direct commands to `wget` and `tar`. Those are not very portable and require use to have a `Dockerfile` that installs `wget`.
As suggested in #6 a better option would be to use the [curl](https://github.com/alexcrichton/curl-rust) crate and [flate2 and tar](https://rust-lang-nursery.github.io/rust-cookbook/compression/tar.html) instead.

I did try that change in the PR mentionned above but faced the following issue when cross compiling:
```
= note: /usr/bin/ld: /target/debug/deps/libopenssl_sys-07dcfde38d575d2d.rlib(s3_clnt.o): Relocations in generic ELF (EM: 183)
          /usr/bin/ld: /target/debug/deps/libopenssl_sys-07dcfde38d575d2d.rlib(s3_clnt.o): Relocations in generic ELF (EM: 183)
          /target/debug/deps/libopenssl_sys-07dcfde38d575d2d.rlib: error adding symbols: File in wrong format
          collect2: error: ld returned 1 exit status
```
I think that is because of clash of OpenSSL libraries between the one on the host and the one used in `cross` Docker container. This needs more investigating.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -8,38 +8,5 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v1
-    - name: Build
-      run: cargo build --verbose
-      # rustfmt will check if the code build before checking the formatting.
-      # Because the build script generated a new module in the code,
-      # building the code before checking it is needed.
-    - name: Check formatting
-      run: cargo fmt --all -- --check
-
-  test:
-    name: Run tests
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v1
-    - run: cargo test --verbose
-
-  linting:
-    name: Execute clippy
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v1
-    - run: cargo clippy --all-targets --all-features -- -D clippy::all -D clippy::cargo
-
-  cross-testing-arm64-linux:
-    name: Cross testing for the aarch64-unknown-linux-gnu target
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v1
-    - name: Install cross
-      run: cargo install cross
-    - name: Build the cross Dockerfile
-      run: docker build -t parsec-cross .
-    - name: Cross-compile with cross
-      run: cross build --target aarch64-unknown-linux-gnu --verbose
-    - name: Execute the unit tests with cross
-      run: cross test --target aarch64-unknown-linux-gnu --verbose
+    - name: Execute all tests
+      run: ./tests/ci.sh

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,0 +1,5 @@
+# Executing our tests on Arm64 with Travis CI
+arch: arm64
+language: rust
+script:
+- ./tests/ci.sh

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "parsec-interface"
-version = "0.6.0"
+version = "0.6.1"
 authors = ["Paul Howard <paul.howard@arm.com>",
            "Ionut Mihalcea <ionut.mihalcea@arm.com>",
            "Hugues de Valon <hugues.devalon@arm.com>"]
@@ -14,6 +14,9 @@ edition = "2018"
 
 [build-dependencies]
 prost-build = "0.5.0"
+curl = "0.4.25"
+flate2 = "1.0.13"
+tar = "0.4.26"
 
 [dependencies]
 serde = { version = "1.0", features = ["derive"] }

--- a/Cross.toml
+++ b/Cross.toml
@@ -1,4 +1,0 @@
-# cross configuration file to use the Dockerfile
-
-[target.aarch64-unknown-linux-gnu]
-image = "parsec-cross"

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,8 +1,0 @@
-# This Dockerfile is used by cross for cross-compilation and cross-testing of
-# Parsec.
-
-FROM rustembedded/cross:aarch64-unknown-linux-gnu-0.1.16
-
-RUN apt-get update && \
-    # wget is needed in the build script to download the operations.
-    apt-get install -y wget

--- a/README.md
+++ b/README.md
@@ -37,6 +37,9 @@ This project uses the following third party crates:
 * uuid (Apache-2.0)
 * log (MIT and Apache-2.0)
 * arbitrary (MIT and Apache-2.0)
+* curl (MIT)
+* flate2 (MIT and Apache-2.0)
+* tar (MIT and Apache-2.0)
 
 # Contributing
 

--- a/build.rs
+++ b/build.rs
@@ -1,4 +1,4 @@
-// Copyright (c) 2019, Arm Limited, All Rights Reserved
+// Copyright (c) 2020, Arm Limited, All Rights Reserved
 // SPDX-License-Identifier: Apache-2.0
 //
 // Licensed under the Apache License, Version 2.0 (the "License"); you may
@@ -12,11 +12,13 @@
 // WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+use curl::easy::Easy;
+use flate2::read::GzDecoder;
 use std::env;
-use std::fs::read_dir;
-use std::io::{Error, ErrorKind, Result};
+use std::fs::{read_dir, File};
+use std::io::{Error, ErrorKind, Result, Write};
 use std::path::Path;
-use std::process::Command;
+use tar::Archive;
 
 const PARSEC_OPERATIONS_VERSION: &str = "0.2.0";
 
@@ -52,36 +54,34 @@ fn generate_proto_sources() -> Result<()> {
 
 fn get_protobuf_files() -> Result<()> {
     // TODO: Use semantic versioning to get the newest versions.
-    if !Command::new("wget")
-        .arg(format!(
-            "https://github.com/parallaxsecond/parsec-operations/archive/{}.tar.gz",
-            PARSEC_OPERATIONS_VERSION
-        ))
-        .arg(format!(
-            "--directory-prefix={}",
-            env::var("OUT_DIR").unwrap()
-        ))
-        .status()?
-        .success()
+    let protobuf_archive_url = format!(
+        "https://codeload.github.com/parallaxsecond/parsec-operations/tar.gz/{}",
+        PARSEC_OPERATIONS_VERSION
+    );
+    let out_dir = env::var("OUT_DIR").unwrap();
+    let protobuf_archive_path = format!("{}/{}.tar.gz", out_dir, PARSEC_OPERATIONS_VERSION);
+    let mut protobuf_archive = File::create(&protobuf_archive_path)?;
+
+    let mut buf = Vec::new();
+    let mut handle = Easy::new();
+    handle.url(&protobuf_archive_url)?;
     {
-        return Err(Error::new(ErrorKind::Other, "wget command failed"));
+        let mut transfer = handle.transfer();
+        transfer.write_function(|data| {
+            buf.extend_from_slice(data);
+            Ok(data.len())
+        })?;
+        transfer.perform()?;
     }
 
-    // Gets extracted as parsec-operations-PARSEC_OPERATIONS_VERSION directory.
-    if !Command::new("tar")
-        .arg("xf")
-        .arg(format!(
-            "{}/{}.tar.gz",
-            env::var("OUT_DIR").unwrap(),
-            PARSEC_OPERATIONS_VERSION
-        ))
-        .arg("--directory")
-        .arg(env::var("OUT_DIR").unwrap())
-        .status()?
-        .success()
-    {
-        return Err(Error::new(ErrorKind::Other, "wget command failed"));
-    }
+    protobuf_archive.write_all(&buf)?;
+
+    // Drop and open the archive again so that GzDecoder sees it as a new file.
+    let protobuf_archive = File::open(&protobuf_archive_path)?;
+
+    let tar = GzDecoder::new(protobuf_archive);
+    let mut archive = Archive::new(tar);
+    archive.unpack(&out_dir)?;
 
     Ok(())
 }

--- a/tests/ci.sh
+++ b/tests/ci.sh
@@ -1,0 +1,48 @@
+#!/usr/bin/env bash
+
+# ------------------------------------------------------------------------------
+# Copyright (c) 2020, Arm Limited, All Rights Reserved
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#          http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------------
+
+# Continuous Integration test script, executed by GitHub Actions on x86 and
+# Travis CI on Arm64.
+
+set -euf -o pipefail
+
+##############
+# Build test #
+##############
+RUST_BACKTRACE=1 cargo build
+
+#################
+# Static checks #
+#################
+# On native target clippy or fmt might not be available.
+if cargo fmt -h
+then
+	cargo fmt --all -- --check
+fi
+if cargo clippy -h
+then
+	cargo clippy --all-targets -- -D clippy::all -D clippy::cargo
+fi
+
+############################
+# Unit tests and doc tests #
+############################
+RUST_BACKTRACE=1 cargo test
+
+cargo clean

